service/history/handler.go

```diff
@@ -2110,51 +2135,74 @@ func (h *handlerImpl) updateErrorMetric(
 	scope metrics.Scope,
 	domainID string,
 	workflowID string,
+	runID string,
 	err error,
 ) {
 
-	if err == context.DeadlineExceeded || err == context.Canceled {
+	var yarpcE *yarpcerrors.Status
+
+	if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
 		scope.IncCounter(metrics.CadenceErrContextTimeoutCounter)
 		return
 	}
 
-	switch err := err.(type) {
-	case *types.ShardOwnershipLostError:
+	if errors.Is(err, types.ShardOwnershipLostError{}) {
 		scope.IncCounter(metrics.CadenceErrShardOwnershipLostCounter)
-	case *types.EventAlreadyStartedError:
+
+	} else if errors.Is(err, types.EventAlreadyStartedError{}) {
 		scope.IncCounter(metrics.CadenceErrEventAlreadyStartedCounter)
-	case *types.BadRequestError:
+
+	} else if errors.Is(err, types.BadRequestError{}) {
 		scope.IncCounter(metrics.CadenceErrBadRequestCounter)
-	case *types.DomainNotActiveError:
+
+	} else if errors.Is(err, types.DomainNotActiveError{}) {
 		scope.IncCounter(metrics.CadenceErrBadRequestCounter)
-	case *types.WorkflowExecutionAlreadyStartedError:
+
+	} else if errors.Is(err, types.WorkflowExecutionAlreadyStartedError{}) {
 		scope.IncCounter(metrics.CadenceErrExecutionAlreadyStartedCounter)
-	case *types.EntityNotExistsError:
+
+	} else if errors.Is(err, types.EntityNotExistsError{}) {
 		scope.IncCounter(metrics.CadenceErrEntityNotExistsCounter)
-	case *types.WorkflowExecutionAlreadyCompletedError:
+
+	} else if errors.Is(err, types.WorkflowExecutionAlreadyCompletedError{}) {
 		scope.IncCounter(metrics.CadenceErrWorkflowExecutionAlreadyCompletedCounter)
-	case *types.CancellationAlreadyRequestedError:
+
+	} else if errors.Is(err, types.CancellationAlreadyRequestedError{}) {
 		scope.IncCounter(metrics.CadenceErrCancellationAlreadyRequestedCounter)
-	case *types.LimitExceededError:
+
+	} else if errors.Is(err, types.LimitExceededError{}) {
 		scope.IncCounter(metrics.CadenceErrLimitExceededCounter)
-	case *types.RetryTaskV2Error:
+
+	} else if errors.Is(err, types.RetryTaskV2Error{}) {
 		scope.IncCounter(metrics.CadenceErrRetryTaskCounter)
-	case *types.ServiceBusyError:
+
+	} else if errors.Is(err, types.ServiceBusyError{}) {
 		scope.IncCounter(metrics.CadenceErrServiceBusyCounter)
-	case *yarpcerrors.Status:
-		if err.Code() == yarpcerrors.CodeDeadlineExceeded {
+
+	} else if errors.As(err, &yarpcE) {
+
+		if yarpcE.Code() == yarpcerrors.CodeDeadlineExceeded {
 			scope.IncCounter(metrics.CadenceErrContextTimeoutCounter)
 		}
 		scope.IncCounter(metrics.CadenceFailures)
```

Comment:
not related to your change: why do we increment both `CadenceErrContextTimeoutCounter` and `CadenceFailures` in this case? if there are dashboards looking at sum of these failures we will see extra numbers.

Reply:
I don't have context there to be honest

Reply:
@Groxx do we have dashboards summing these up and showing as total error count?